refactor(storage)!: add a common blobstore to store all blobs (#4136) (#4137)
project-zot/zot: pkg/storage/README.md

This document reflects the finalized storage design decisions for proposal #3750.

zot supports two classes of storage backends:

1. local filesystems
2. remote object stores (for example S3, GCS, Azure)

The cache database is configured independently of the blob backend.

> ## Dedupe Design (Final)

Reviewer: Since using git, "Final" isn't something you'd want in the code base :)

This repository uses a single global blob namespace named `_blobstore` for both local and remote dedupe flows.

1. Blob content is promoted to `_blobstore/blobs/<algorithm>/<digest>`.
2. Repository blob paths keep per-repo ownership semantics.
3. On remote backends, per-repo paths are marker objects and reads resolve deterministically from `_blobstore`.
4. On local filesystems, dedupe still relies on hardlinks.

> ## Migration Behavior (Final)

Reviewer: Since using git, "Final" isn't something you'd want in the code base :)

Legacy layouts are upgraded automatically at startup when dedupe is enabled.

1. Migration is marker-guarded by `_global_blobstore_migrated` at the image store root.
2. If the marker exists, startup skips migration.
3. If migration is incomplete, startup retries on the next launch.
4. There is no user-facing migrate or rollback CLI for this flow.

## Downgrade Policy

Remote downgrade across this dedupe migration is unsupported.

After migration to `_blobstore` marker-based remote layout, running an older remote dedupe layout is not a supported path.

## Migration Matrix

| Backend | Direction | Support | Notes |
| --- | --- | --- | --- |
| local filesystem | legacy per-repo blobs -> `_blobstore` layout | supported | Automatic at startup when dedupe is enabled; migration marker prevents repeated full scans. |

> | local filesystem | `_blobstore` layout -> older local release | conditionally supported | No dedicated rollback flow; behavior depends on older release expectations and filesystem hardlink semantics. |

Reviewer: Do we want to claim support for this flow? IMO, if not, it would be safer to set up 2 zots and sync the content to the old one from the new if someone wants to go from new -> old.

Reviewer 2: I agree, why do we mention "conditionally supported". That is very vague.

What does conditionally refer to? "depends on older release expectations" we know older release expectations of storage layout and they are not the same as this release.

From my understanding, "supported" means that if there's any bug with switching from new zot to old zot, we promise to fix it in zot code, which is not something we intend to do.

| remote object store (S3/GCS/Azure) | legacy per-repo blobs -> `_blobstore` + marker layout | supported | Automatic at startup when dedupe is enabled; marker-guarded and resumable on next startup if incomplete. |

> | remote object store (S3/GCS/Azure) | `_blobstore` + marker layout -> older remote release | unsupported | No rollback CLI is provided; remote downgrade is not a supported compatibility path. |

Reviewer 2: ```suggestion
| remote object store (S3/GCS/Azure) | `_blobstore` + marker layout -> older remote release | unsupported | Remote downgrade is not a supported compatibility path. |
```

We already mention there's no rollback CLI in another paragraph.

> ## Cache Backends

Reviewer 2: This section is a bit off. Doesn't add anything of value. Doesn't even explain what a `Cache Backend` is or does.

> This document reflects the storage design decisions for proposal #3750.

Reviewer: IMO, this document should be titled "Storage Subsystem Design" and this line could be removed.

zot supports two classes of storage backends:

1. local filesystems
2. remote object stores (for example S3, GCS, Azure)

> The cache database is configured independently of the blob backend.

Reviewer: ```suggestion
The storage backend stores repos, artifact blobs, and referrers. The cache database is configured independently of the storage backend.
```

> ## Dedupe Design

Reviewer: Before getting into dedupe, I think there should be 2 sections - one for local and one for remote.
Each section should describe the lifecycle of a blob under that config.

This repository uses a single global blob namespace named `_blobstore` for both local and remote dedupe flows.

1. Blob content is promoted to `_blobstore/blobs/<algorithm>/<digest>`.

> 2. Repository blob paths keep per-repo ownership semantics.

Reviewer: This line should add more details on how the repo paths maintain the per-repo ownership semantic.

> 3. On remote backends, per-repo paths are marker objects and reads resolve deterministically from `_blobstore`.

Reviewer: This line should add more details on how the deterministic resolution works.

4. On local filesystems, dedupe still relies on hardlinks.

## Migration Behavior

> Legacy layouts are upgraded automatically at startup when dedupe is enabled.

Reviewer: It would help to mention at which stage of startup, this upgrade takes place (before HTTP server, but after ..., etc.)

1. Migration is marker-guarded by `_global_blobstore_migrated` at the image store root.
2. If the marker exists, startup skips migration.
3. If migration is incomplete, startup retries on the next launch.
4. There is no user-facing migrate or rollback CLI for this flow.

## Downgrade Policy

Downgrade across this dedupe migration is unsupported for both local and remote backends.

> After migration to the `_blobstore` layout, running an older release against that same storage path is not a supported path, regardless of backend. There is no dedicated rollback flow, and behavior would depend on the older release's expectations and (for local filesystems) hardlink semantics. If you need to go from a new zot back to an older one, run a separate zot instance on the old layout and sync content into it from the new instance rather than downgrading in place.

Reviewer: ```suggestion
After migration to the `_blobstore` layout, running an older release against that same storage path is not a supported path, regardless of backend. There is no dedicated rollback flow to a release that does not have the new storage system. If you need to go from a new zot back to an older one, run a separate zot instance on the old layout and sync content into it from the new instance rather than downgrading in place.
```
